ratpack-test/src/main/java/ratpack/test/mock/MockApi.java

```diff
@@ -0,0 +1,145 @@
+
+/**
+ * A test harness for simulating behavior of remote APIs by starting an {@link EmbeddedApp} that will
+ * handle requests based on the content of the received request.
+ * <pre class="java">{@code
+ * import ratpack.http.HttpMethod;
+ * import ratpack.http.client.HttpClient;
+ * import ratpack.test.embed.EmbeddedApp;
+ * import ratpack.test.handling.HandlerFactory;
+ * import ratpack.test.mock.MockApi;
+ *
+ * import static org.junit.Assert.assertEquals;
+ *
+ * public class Example {
+ *   public static void main(String... args) throws Exception {
+ *
+ *     HandlerFactory factory = request -> {
+ *       if (request.getMethod() ==  HttpMethod.GET)  {
+ *         return ctx -> ctx.render("get on remote API");
+ *       }
+ *       return ctx -> ctx.getResponse().status(400).send();
+ *     };
+ *
+ *     MockApi remoteApi = MockApi.of(factory);
+ *
+ *     EmbeddedApp.fromHandlers(chain -> {
+ *       chain.get("get", ctx ->
+ *         ctx.get(HttpClient.class)
+ *           .get(remoteApi.getAddress())
+ *           .then(resp ->
+ *             resp.forwardTo(ctx.getResponse())
+ *           )
+ *       );
+ *       chain.get("post", ctx ->
+ *         ctx.get(HttpClient.class)
+ *           .post(remoteApi.getAddress(), spec -> {})
+ *           .then(resp ->
+ *             resp.forwardTo(ctx.getResponse())
+ *           )
+ *       );
+ *     }).test(httpClient -> {
+ *       assertEquals("get on remote API", httpClient.get("get").getBody().getText());
+ *       assertEquals(400, httpClient.get("post").getStatusCode());
+ *     });
+ *   }
+ * }
+ * }</pre>
+ * <p>
+ * This fixture is particularly powerful when combined with
+ * Spock's {@code Mock} API by providing a Mock {@link HandlerFactory}
+ * to this class. Iteractions to a remote API can then be
+ * validated inline to the {@code Specification} by
+ * verifying the invocations of the mock {@code HandlerFactory}.
+ * <pre class="groovy">{@code
+ * import ratpack.groovy.test.embed.GroovyEmbeddedApp
+ * import ratpack.http.HttpMethod
+ * import ratpack.http.client.HttpClient
+ * import spock.lang.Specification
+ *
+ * import static ratpack.groovy.Groovy.groovyHandler
+ *
+ * class ApiSpec extends Specification {
+ *
+ *   def "test api"() {
+ *     given:
+ *     MockApi remoteApi = MockApi.of(Mock(HandlerFactory))
+ *
+ *     def app = GroovyEmbeddedApp.ratpack {
+ *       handlers {
+ *         get { ctx ->
+ *           ctx.get(HttpClient).get(remoteApi.address).then { resp ->
+ *             resp.forwardTo(ctx.response)
+ *           }
+ *         }
+ *       }
+ *     }
+ *
+ *     when:
+ *     def resp = app.httpClient.get()
+ *
+ *     then:
+ *     1 * remoteApi.handlerFactory.receive({
+ *       it.method == HttMethod.GET
+ *       it.path == ""
+ *     } >> groovyHandler {
+ *       render("remote ok")
+ *     }
+ *     resp.body.text == "remote ok"
+ *   }
+ * }
+ * }</pre>
+ * @since 1.7.0
+ */
+public class MockApi implements EmbeddedApp {
```

Comment:
This is cool - can you add documentation and/or convenience method to mock out a chain of requests? We often have multitple endpoints to mock together. Something like:

```
HandlerFactory factory = request -> Handlers.chain(request, chain -> chain.get()); 
MockApi mockApi = MockApi.of(request);
```

Maybe something like:
```
public static HandlerFactory of(Action<? super Chain> action) {
    return request -> Handlers.chain(request, action);
}
```

Reply:
well, the intent here is to create a test fixture that you can combine both spock mock invocation and end-to-end http requests, with this you can excercise a full AUT talking to embedded apps over real network, but easily configure how the embedded apps handle each request inline to a spec.

Its not intended as a replacement for wiremock as that does offer some lower level controls for testing things like connection resets and such, but as a mocking response behavior, this maybe be a little cleaner.

Reply:
I don't see how providing a chain would work, in that case why not just make an embedded app?

Reply:
@johnrengelman makes sense - this feature will functionally doing similar things that WireMock also does, via a different and more lightweight mechanism. Perhaps calling out when to use this vs. something heavier like WireMock or BetaMax would be useful to end users.